core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java

```diff
@@ -1054,10 +1053,6 @@ private RelNode changeTraitsUsingConverters(
               allowInfiniteCostConverters);
       if (rel != null) {
         assert rel.getTraitSet().getTrait(traitDef).satisfies(toTrait);
-        rel =
-            completeConversion(
-                rel, allowInfiniteCostConverters, toTraits,
-                Expressions.list(traitDef));
         if (rel != null) {
           register(rel, converted);
         }
```

Comment:
https://github.com/apache/calcite/pull/1586/files#diff-70d5035797baebe157bab8cbf860e456R1061 can be changed to `else if` and remove `rel == null` check.